### PR TITLE
Improve backtrace looks and performance

### LINKED ISSUE #247: Inexplicable hang on Emacs 30 snapshot
Hi,

I'm seeing that my tests for org-ql with Buttercup are hanging exclusively on Emacs snapshot (currently an Emacs 30 pre-release version).  I don't know if this indicates a bug in Buttercup, or an incompatible change in Emacs 30, but I hope that we can find the problem and solve it before Emacs 30.1 is released.  These hangs are happening with the latest version of Buttercup installed automatically from MELPA during the CI run.

Perhaps notably, perhaps related, Guix has disabled tests for org-ql due to what seems to be a similar problem: https://git.savannah.gnu.org/cgit/guix.git/commit/gnu/packages/emacs-xyz.scm?id=3add97c7761e6c58a1d7405f417a49dda5f0a742

```
* gnu/packages/emacs-xyz.scm (emacs-org-ql): Update to 0.8.6.
[arguments]<#:tests?>: Disable tests, which freeze starting with recent
Buttercup releases.
```

I can't reproduce the problem locally, being on a different Emacs version.  But besides that, earlier today while debugging similar issues, I had similar hangs during a Buttercup run on other Emacs/Org/Buttercup versions, and I don't know how to "look inside" a Buttercup run to find out what's hanging.

Any suggestions would be appreciated.  Thanks.

### PR DESCRIPTION
**Pull request by snogge**
Make the backtrace prints better - less unwanted output - with a lot better performance.
Fixes #247 and #220.

### CODE REVIEW

**Comment by snogge** [comment]
@alphapapa @doublep @jcs090218 
I did some major surgery on the backtrace code which really help with performance. 
I'll give you a few days to check it out before I merge. Let me know if I broke anything.

**Comment by alphapapa** [comment]
Hi Ola,

Thanks for the heads-up.  I just did a quick test on the `org-ql` repo, both with all tests passing and with 104 out of 437 specs failing, and performance was excellent: instead of taking minutes to finish outputting the results and the backtraces, it happened in a matter of seconds.  This is a great improvement.  Thank you for working on it!

**Comment by jcs090218** [comment]
Thank you for notifying me. I've done a quick test, and it is working fine. :)

I've opened up a PR in #250 to test downstream Eask and see if it's working. This could make you more confident about the changes. 🤔

**Comment by snogge** [comment]
@doublep , how can I recreate that?

**Comment by doublep** [comment]
~/git/buttercup$ git diff
    diff --git a/tests/test-buttercup.el b/tests/test-buttercup.el
    index 5172c46..02420b3 100644
    --- a/tests/test-buttercup.el
    +++ b/tests/test-buttercup.el
    @@ -172,7 +172,7 @@ before it's processed by other functions."
       (describe "should handle"
         (it "expressions wrapped by buttercup--wrap-expr"
           (expect (buttercup--enclosed-expr (buttercup--wrap-expr-and-eval '(ignore)))
    -              :to-equal '(ignore)))
    +              :to-equal '(INTENTIONALLY-BROKEN-TEST)))
         (it "a closure with expression copy?"
           ;; This is for before Oclosures were added, and is not testable
           ;; once interpreted-function types were added in Emacs 30.
    ~/git/buttercup$ emacs --version
    GNU Emacs 29.4.50
    Development version c0ef9e766eb1 on emacs-29 branch; build date 2024-06-28.
    Copyright (C) 2024 Free Software Foundation, Inc.
    GNU Emacs comes with ABSOLUTELY NO WARRANTY.
    You may redistribute copies of GNU Emacs
    under the terms of the GNU General Public License.
    For more information about these matters, see the file named COPYING.
    ~/git/buttercup$ make test
    ./bin/buttercup -L . tests 
    [...]

    Traceback (most recent call last):
      normal-top-level()
      command-line()
      command-line-1(("-L" "." "-l" "buttercup" "-f" "buttercup-run-discover" "-...
      buttercup-run-discover()
      buttercup-run()
      buttercup--run-suites((#s(buttercup-suite "The buttercup--enclosed-expr fu...
      mapc(buttercup--run-suite (#s(buttercup-suite "The buttercup--enclosed-exp...
      buttercup--run-suite(#s(buttercup-suite "The buttercup--enclosed-expr func...
      buttercup--run-suite(#s(buttercup-suite "should handle" #s(buttercup-suite...
      buttercup--run-spec(#s(buttercup-spec "expressions wrapped by buttercup--w...
      buttercup--update-with-funcall(#s(buttercup-spec "expressions wrapped by b...
      apply(buttercup--funcall (closure (t) nil (condition-case err (progn (butt...
      buttercup--funcall((closure (t) nil (condition-case err (progn (buttercup-...
      apply((closure (t) nil (condition-case err (progn (buttercup-expect (let (...
      (closure (t) nil (condition-case err (progn (buttercup-expect (let ((temp0...
      (condition-case err (progn (buttercup-expect (let ((temp0 '(buttercup--enc...
      (progn (buttercup-expect (let ((temp0 '(buttercup--enclosed-expr (buttercu...
      buttercup-expect((closure ((:type . buttercup--thunk) (expr buttercup--enc...
      buttercup-fail("%s" "Expected `(buttercup--enclosed-expr (buttercup--wrap-...
      signal(buttercup-failed "Expected `(buttercup--enclosed-expr (buttercup--w...
    FAILED: Expected `(buttercup--enclosed-expr (buttercup--wrap-expr-and-eval '(ignore)))' to be `equal' to `(INTENTIONALLY-BROKEN-TEST)', but instead it was `(ignore)' which does not match because: (list-elt 0 (different-atoms ignore INTENTIONALLY-BROKEN-TEST)).

    Ran 283 out of 284 specs, 1 failed, in 138.10ms.
    buttercup-run failed: ""
    make: *** [Makefile:12: check-buttercup] Error 255

**Comment by snogge** [comment]
Hmm, but that specific test is not nested.  I'll have to look further.

**Comment by doublep** [comment]
> If you can recreate it with another package I would really like to know about it.

I originally noticed it with `test/project-e` in Eldev's test suite. I did run Buttercup inside Eldev, so initially I suspected that something got broken in my integration code because of these changes. Then I retested it without Eldev at all, on Buttercup directly, and saw the same problem. So, I'm pretty sure that it's a problem with this branch and most likely it's not about nesting level, because it seems to happen both in Buttercup and in unrelated projects.

**Comment by snogge** [comment]
I think I'm on to something now.
The new code works great for unexpected errors in code called from *within* an `except` statement.  That's why it works for the failing `org-ql` tests - there was a bug in `org-ql` that caused `invalid-function` errors. 
Code outside `except` is not handled well on this branch.
The problem with the failing test in buttercup is actually different. That is just a test-mismatch and should really not collect any backtrace at all. What's a meaningful backtrace for `(expect 1 :to-equal 0)`?
There are a few other corner cases that have to be handled as well, like errors in the matcher or in before/after blocks.

**Comment by snogge** [comment]
Newer and even more improved. 
Now also with tests.
@doublep, the problems you reported should be fixed now.
@jcs090218 , the tests #250 passes on this branch.

**Comment by snogge** [comment]
No comments for more than a week, merging this now.
